pkg/controller/remotemachineset/gcpactuator.go

```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
 		return "", errors.New(msg)
 	}
 }
+
+// obtainLease uses the Hive MachinePoolNameLease resource to obtain a unique, single character
+// for use in the name of the machine pool. We are severely restricted on name lengths on GCP
+// and effectively have one character of flexibility with the naming convention originating in
+// the installer.
+func (a *GCPActuator) obtainLease(pool *hivev1.MachinePool, logger log.FieldLogger) (rune, error) {
+	logger.Debugf("obtaining lease for pool: %s", pool.Name)
+
+	cd := &hivev1.ClusterDeployment{}
+	err := a.client.Get(context.TODO(), types.NamespacedName{Namespace: pool.Namespace, Name: pool.Spec.ClusterDeploymentRef.Name}, cd)
+	if err != nil {
+		return 'X', err
+	}
+
+	leases := &hivev1.MachinePoolNameLeaseList{}
+	// TODO: pass ^^ to find command?
+	err = a.client.List(context.TODO(), leases, client.UseListOptions(&client.ListOptions{Namespace: pool.Namespace}))
+	if err != nil {
+		return 'X', err
+	}
+	logger.Debugf("found %d leases in namespace", len(leases.Items))
+	for _, l := range leases.Items {
+		for _, ownerRef := range l.OwnerReferences {
+			logger.Debugf("owner ref: %v", ownerRef)
+			if ownerRef.APIVersion == "hive.openshift.io/v1" &&
+				ownerRef.Kind == "MachinePool" &&
+				ownerRef.Name == pool.Name {
+				logger.Info("machine pool already has lease: %s", l.Name)
+				return rune(l.Name[len(l.Name)-1]), nil
+			}
+		}
+	}
+
+	// Pool does not have a lease yet, lookup all currently available lease chars
+	availLeaseChars, err := a.findAvailableLeaseChars(cd)
```

Comment:
Why not pass in the `MachinePoolNameLeaseList` to `findAvailableLeaseChars` rather than listing again?